File: lab4/C++/main.cpp
    #include <iostream>
    #include <vector>
    #include "Desenvolvedor.hpp"
    #include "Gerente.hpp"
    #include "TechLead.hpp"
    using namespace std;

    int main()
    {
        vector<Desenvolvedor> desenvolvedores;
        vector<Gerente> gerentes;
        vector<TechLead> techLeads;
        int op = 0;
        do
        {
            cout << "\nMenu: " << endl;
            cout << "1--->Adicionar Desenvolvedor. " << endl;
            cout << "2--->Adicionar Gerente. " << endl;
            cout << "3--->Adicionar TechLead. " << endl;
            cout << "4--->Imprimir funcionários cadastrados. " << endl;
            cout << "0--->Sair" << endl;
            cout << "Digite qual opção desejada: ";
            cin >> op;
            cin.ignore();
            cout << endl;

            string nomeGerente, nomeDesenvolvedor, linguagemPrincipal, nomeTechLead;
            int qtdEquipes = 0, qtdProjetos = 0;
            float salarioG = 0.0, salarioD = 0.0, salarioTL = 0.0;

            if(op == 1){
                cout << "Digite o nome do desenvolvedor: ";
                getline(cin, nomeDesenvolvedor);
                cout << "Digite seu salário: ";
                cin >> salarioD;
                cin.ignore();
                cout << "Digite a linguagem principal: ";
                getline(cin, linguagemPrincipal);
                Desenvolvedor desenvolvedor(nomeDesenvolvedor, salarioD, linguagemPrincipal);
                desenvolvedores.push_back(desenvolvedor);
            }
            else if (op == 2){ // add gerente
                cout << "Digite o nome do gerente: ";
                getline(cin, nomeGerente);
                cout << "Digite seu salário: ";
                cin >> salarioG;
                cin.ignore();
                cout << "Digite a quantidade de equipes sob sua supervisão: ";
                cin >> qtdEquipes;
                cin.ignore();
                Gerente gerente(nomeGerente, salarioG, qtdEquipes);
                gerentes.push_back(gerente);
            }
            else if (op == 3){ // add techLead
                cout << "Digite o nome do TechLead: ";
                getline(cin, nomeTechLead);
                cout << "Digite seu salário: ";
                cin >> salarioTL;
                cin.ignore();
                cout << "Digite a quantidade de projetos em sua direção: ";
                cin >> qtdEquipes;
                cin.ignore();
                cout << "Digite a linguagem principal: ";
                getline(cin, linguagemPrincipal);
                cout << "Digite a quantidade projetos realizados: ";
                cin >> qtdProjetos;
                
                TechLead techLead(nomeTechLead, salarioTL, qtdEquipes, linguagemPrincipal, qtdProjetos);
                techLeads.push_back(techLead);
            }
            else if (op == 4){ // imprimir
                for(auto& dev : desenvolvedores) 
                    cout << dev.descricao() << endl;
                for(auto& gerente : gerentes) 
                    cout << gerente.descricao() << endl;
                for(auto& techLead : techLeads) 
                    cout << techLead.descricao() << endl;
            }
            else if (op == 0){
                cout << "Saindo..." << endl;
                break;
            }
            else{
                cout << "Opção " << op << " inválida. " << endl;
            }
        } while (op != 0);
        return 0;
    }

File: lab4/Java/Main.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        // Listas para armazenar os funcionários
        List<Desenvolvedor> desenvolvedores = new ArrayList<>();
        List<Gerente> gerentes = new ArrayList<>();
        List<TechLead> techLeads = new ArrayList<>();

        // Scanner para entrada de dados
        Scanner scanner = new Scanner(System.in);
        int op = -1;

        do {
            // Menu de opções
            System.out.println("\nMenu: ");
            System.out.println("1 ---> Adicionar Desenvolvedor.");
            System.out.println("2 ---> Adicionar Gerente.");
            System.out.println("3 ---> Adicionar TechLead.");
            System.out.println("4 ---> Imprimir funcionários cadastrados.");
            System.out.println("0 ---> Sair");
            System.out.print("Digite qual opção desejada: ");

            op = scanner.nextInt();
            scanner.nextLine();
            
            String nomeDesenvolvedor, linguagemPrincipal, nomeGerente, nomeTechLead;
            int qtdEquipes = 0, qtdProjetos = 0;
            float salarioD = 0, salarioG = 0, salarioTL = 0;

            switch (op) {
                case 1: // add dev
                    System.out.print("Digite o nome do desenvolvedor: ");
                    nomeDesenvolvedor = scanner.nextLine();
                    
                    System.out.print("Digite seu salário: ");
                    salarioD = scanner.nextFloat();
                    scanner.nextLine();
                    
                    System.out.print("Digite a linguagem principal: ");
                    linguagemPrincipal = scanner.nextLine();
                    System.out.println("Digite a quantidade de  projetos reslizados?");
                    int qtdProjetosRealizados = scanner.nextInt();
                    scanner.nextLine();
                    
                    Desenvolvedor desenvolvedor = new Desenvolvedor(nomeDesenvolvedor, salarioD, linguagemPrincipal, qtdProjetosRealizados);
                    desenvolvedores.add(desenvolvedor);
                    System.out.println("Desenvolvedor adicionado com sucesso!");
                    break;

                case 2: //add gerente
                    System.out.print("Digite o nome do gerente: ");
                    nomeGerente = scanner.nextLine();
                    
                    System.out.print("Digite seu salário: ");
                    salarioG = scanner.nextFloat();
                    scanner.nextLine();

                    System.out.print("Digite a quantidade de equipes sob sua supervisão: ");
                    qtdEquipes = Integer.parseInt(scanner.nextLine());
                    
                    Gerente gerente = new Gerente(nomeGerente, salarioG, qtdEquipes);
                    gerentes.add(gerente);
                    System.out.println("Gerente adicionado com sucesso!");
                    break;

                case 3: // add techlead
                    System.out.print("Digite o nome do TechLead: ");
                    nomeTechLead = scanner.nextLine();
                    
                    System.out.print("Digite seu salário: ");
                    salarioTL = scanner.nextFloat();
                    scanner.nextLine();
                    
                    System.out.print("Digite a quantidade de equipes sob sua supervisão: ");
                    qtdEquipes = Integer.parseInt(scanner.nextLine());
                    
                    System.out.print("Digite a linguagem principal: ");
                    linguagemPrincipal = scanner.nextLine();
                    
                    System.out.print("Digite a quantidade de projetos em andamento: ");
                    qtdProjetos = scanner.nextInt();
                    scanner.nextLine();
                    
                    TechLead techLead = new TechLead(nomeTechLead, salarioTL, qtdEquipes, linguagemPrincipal, qtdProjetos);
                    techLeads.add(techLead);
                    System.out.println("TechLead adicionado com sucesso!");
                    break;

                case 4: // Imprimir funcionários
                    System.out.println("\n--- Desenvolvedores ---");
                    for (Desenvolvedor dev : desenvolvedores) {
                        System.out.println(dev.descricao());
                        System.out.println(dev.projetosRealizados());
                    }

                    System.out.println("\n--- Gerentes ---");
                    for (Gerente ger : gerentes) {
                        System.out.println(ger.descricao());
                        System.out.printf("Bônus: %.2f\n", ger.bonus());
                    }

                    System.out.println("\n--- TechLeads ---");
                    for (TechLead tl : techLeads) {
                        System.out.println(tl.descricao());
                        System.out.printf("Bônus: %.2f\n", tl.bonus());
                        System.out.println(tl.projetosRealizados());
                    }
                    break;

                case 0:
                    System.out.println("Saindo...");
                    break;

                default:
                    System.out.println("Opção " + op + " inválida.");
            }
        } while (op != 0);

        scanner.close();
    }
}